supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigAsyncJobExecutionActor.scala

```diff
@@ -248,44 +251,84 @@ class DispatchedConfigAsyncJobExecutionActor(override val standardParams: Standa
-        jobLogger.error(s"Return file not found after ${exitCodeTimeout.getOrElse("-")} seconds, assuming external kill")
-        val writer = new PrintWriter(jobPaths.returnCode.toFile)
-        // 137 does mean a external kill -9, this is a assumption but easy workaround for now
-        writer.println(9)
-        writer.close()
-        SharedFileSystemRunState("Failed")
-      case (Some(s), _) => s
+        SharedFileSystemJobDone
+
+      case Some(running: SharedFileSystemJobRunning) =>
+        if (running.stale) {
+          // Since the exit code timeout has expired for a running job, check whether the job is still alive.
+
+          isAlive(handle.pendingJob) match {
+            case Success(true) =>
+              // The job is still running. Don't check again for another timeout.
+              SharedFileSystemJobRunning(nextTimeout)
+            case Success(false) =>
+              // The job has stopped but we don't have an RC yet. We'll wait one more 'timeout' for the RC to arrive:
+              SharedFileSystemJobWaitingForReturnCode(nextTimeout)
+            case Failure(e) =>
+              log.error(e, s"Failed to check status for ${handle.jobDescriptor.key.tag} using command: ${checkAliveArgs(handle.pendingJob)}")
+              SharedFileSystemJobRunning(nextTimeout)
+          }
+        } else {
+          // Not stale yet so keep on running!
+          running
+        }
+
+      case Some(waiting: SharedFileSystemJobWaitingForReturnCode) =>
+        if (waiting.giveUpWaiting) {
+          // Can only enter this state when the exit code does not exist and the job is not alive anymore
+          // `isAlive` is not called anymore from this point
+
+          // If exit-code-timeout is set in the config cromwell will create a fake exitcode file
+          val backupError = "??? (!! Programmer Error: It should be impossible to give up on 'waiting' without having set a maximum wait timeout. Please report this as a bug in the Cromwell Github repository !!)"
+          jobLogger.error(s"Return file not found after ${exitCodeTimeout.getOrElse(backupError)} seconds, assuming external kill")
+
+          val returnCodeTemp = jobPaths.returnCode.plusExt("kill")
+          returnCodeTemp.write(s"$SIGTERM${System.lineSeparator}")
+          try {
+            returnCodeTemp.moveTo(jobPaths.returnCode)
+            SharedFileSystemJobFailed
+          } catch {
+            case _: FileAlreadyExistsException =>
+              // If the process has already completed, there will be an existing RC:
+              // Essentially, this covers the rare race condition whereby the task completes between starting to write the
+              // fake RC, and trying to copy it:
+              returnCodeTemp.delete(true)
+              // Looks like a new RC file has appeared - let's go back out of this loop and wait for it:
+              waiting
```

Comment:
If the task completed, why are we waiting?

Is it that Cromwell needs to hold on for another clock tick before it picks up the real RC?

Reply:
It's more like "Cromwell should pick this up on the next poll anyway, so there's no need to reimplement the same logic again here". Sort of like "recurse back into the same function again", albeit on a delay

I guess I should add a warning log in case it gets wedged here somehow.